fw/http_parser.c

```diff
@@ -12583,3 +12583,36 @@ tfw_http_parse_is_done(TfwHttpMsg *hm)
 {
 	return test_bit(TFW_HTTP_B_FULLY_PARSED, hm->flags);
 }
+
+void
+tfw_http_set_req_host(TfwHttpReq *req)
+{
+	int hid = 0;
+	TfwStr *hdrs = req->h_tbl->tbl;
+
+	if (TFW_MSG_H2(req)) {
+		/* RFC 9113, sec-8.3.1:
+		 * The recipient of an HTTP/2 request MUST NOT use
+		 * the Host header field to determine the target
+		 * URI if ":authority" is present.*/
+		if (!TFW_STR_EMPTY(&hdrs[TFW_HTTP_HDR_H2_AUTHORITY]))
+			hid = TFW_HTTP_HDR_H2_AUTHORITY;
+		else
+			hid = TFW_HTTP_HDR_HOST;
+		__h2_msg_hdr_val(&hdrs[hid], &req->host);
+	} else {
+		/* req->host can be only filled by HTTP/1.x parser from
+		 * absoluteURI, so we act as described by RFC 9112, sec-3.2.2
+		 * (https://www.rfc-editor.org/rfc/rfc9112.html#section-3.2.2):
+		 * When an origin server receives a request with an
+		 * absolute-form of request-target, the origin server
+		 * MUST ignore the received Host header field (if any)
+		 * and instead use the host information of the request-target.
+		 */
+		if (TFW_STR_EMPTY(&req->host)) {
+			tfw_http_msg_clnthdr_val(req, &hdrs[TFW_HTTP_HDR_HOST],
+						 TFW_HTTP_HDR_HOST,
+						 &req->host);
+		}
+	}
+}
```

Comment:
Fixed. Copy this function to tests. I think about it, but this decision has one big disadvantage: in case when we change this function in tempesta we can forget to do it in tests and they become incorrect.

Reply:
Yeah, agree, but I hope the tests will fail and we fix it.